druid/src/command.rs

```diff
@@ -15,19 +15,69 @@
 //! Custom commands.
 
 use std::any::Any;
-use std::sync::{Arc, Mutex};
+use std::{
+    marker::PhantomData,
+    sync::{Arc, Mutex},
+};
 
 use crate::{WidgetId, WindowId};
 
+/// An untyped identifier for a `Selector`.
+pub type SelectorSymbol = &'static str;
+
+/// An identifier for a particular command.
+///
+/// This should be a unique string identifier. Certain `Selector`s are defined
+/// by druid, and have special meaning to the framework; these are listed in the
+/// [`druid::commands`] module.
+///
+/// [`druid::commands`]: commands/index.html
+#[derive(Debug, PartialEq, Eq)]
+pub struct Selector<T = ()>(SelectorSymbol, PhantomData<*const T>);
+
+// This has do be done explicitly, to avoid the Copy bound on `T`.
+// See https://doc.rust-lang.org/std/marker/trait.Copy.html#how-can-i-implement-copy .
+impl<T> Copy for Selector<T> {}
+impl<T> Clone for Selector<T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
 /// An identifier for a particular command.
 ///
 /// This should be a unique string identifier. Certain `Selector`s are defined
 /// by druid, and have special meaning to the framework; these are listed in the
 /// [`druid::commands`] module.
 ///
 /// [`druid::commands`]: commands/index.html
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct Selector(&'static str);
+#[derive(Debug, PartialEq, Eq)]
+pub struct OneShotSelector<T>(SelectorSymbol, PhantomData<*const T>);
```

Comment:
I'm not totally sure about this; what it comes down to is whether or not we expect external people to need to use this, or whether it's an implementation detail used in a few places inside druid. Again, I think that API simplicity is important, and I think the more concepts we introduce the more work we need to do to explain what they're for and how to use them. My _feeling_ is that 95+% of the time, the user doesn't need to know about one-shot commands?

A kind of gross way we could encode one-shotness would be in the selector symbol itself, like we could have some keyword that prefaced one-shot commands...

Reply:
Well the ideas I had where:

 Leave it as it is.

- IMO the worst choice as it makes it less obvious how a command is supposed to be used, and would require you to look up every commands doc in order to find out how the hell you are supposed to submit it and how to retrieve the object. Also `Command::get` and `Command::take` will either have to panic or require an additional error, which would add to the interface at least as much noise as a new type.

 `Selector(key: &str, is_once: bool)` and a `Selector::once` constructor.

- Not much better than the previous one.

Encoding it in the selector symbol.

- Again, does not add anything really.

Having a separate type for it.

- Allows `Command::get` to return `Option` instead of `Result`.
- Submitting and retrieving becomes a no-brainer as the compiler checks it for you.
- Could allow merging `Command::new` and `Command::once` if we want to.
- No need to look up a selectors docs to know how to use it (unless it does something very special)
- Those who do not care about one shot commands do not have to use this type

Which is why I very much like this design.

Reply:
I wonder if another option is to encode it in the type param? Like have some type `SingleUse<T>`, and then you're required to explicitly `take` from that... I don't know if I like this, just spitballing.